reactor-core/src/main/java/reactor/core/publisher/FluxWindowTimeout.java

```diff
@@ -26,52 +26,1590 @@
+								return;
+							}
+
+							if (!hasWorkInProgress(expectedState)) {
+								return;
+							}
+						}
+					} else {
+						final InnerWindow<T> nextWindow =
+								new InnerWindow<>(this.maxSize, this, nextWindowIndex, false);
+
+						final InnerWindow<T> previousWindow = this.window;
+
+						this.window = nextWindow;
+
+						this.actual.onNext(nextWindow);
+
+						if (n != Long.MAX_VALUE) {
+							REQUESTED.decrementAndGet(this);
+						}
+
+						previousState = commitWork(this, expectedState, false);
+						expectedState =
+								(((previousState &~ACTIVE_WINDOW_INDEX_MASK) &~ HAS_UNSENT_WINDOW) ^ (expectedState == previousState ? HAS_WORK_IN_PROGRESS : 0)) |
+										incrementActiveWindowIndex(previousState);
+
+						if (isCancelled(expectedState)) {
+							previousWindow.sendCancel();
+							nextWindow.sendCancel();
+							return;
+						}
+
+						if (isTerminated(expectedState)) {
+							final Throwable e = this.error;
+							if (e != null) {
+								previousWindow.sendError(e);
+								nextWindow.sendError(e);
+								this.actual.onError(e);
+							}
+							else {
+								previousWindow.sendComplete();
+								nextWindow.sendComplete();
+								this.actual.onComplete();
+							}
+							return;
+						}
+
+						try {
+							nextWindow.scheduleTimeout();
+						}
+						catch (Exception e) {
+							if (hasWorkInProgress(expectedState)) {
+								this.actual.onError(Operators.onOperatorError(this.s, e, this.actual.currentContext()));
+							}
+							else {
+								this.onError(Operators.onOperatorError(this.s, e, this.actual.currentContext()));
+							}
+
+							return;
+						}
+
+						final long nextRequest;
+						if (previousWindow == null) {
+							// possible at the very beginning
+							nextRequest = this.maxSize;
+						}
+						else {
+							long previousActiveWindowState = previousWindow.sendComplete();
+							nextRequest = InnerWindow.received(previousActiveWindowState);
+						}
+
+						if (nextRequest > 0) {
+							this.s.request(nextRequest);
+						}
+
+						if (!hasWorkInProgress(expectedState)) {
+							return;
+						}
+					}
+				}
+				else if (n == 0 && !hasUnsentWindow) {
+					final InnerWindow<T> nextWindow =
+							new InnerWindow<>(this.maxSize, this, nextWindowIndex, true);
+
+					final InnerWindow<T> previousWindow = this.window;
+					this.window = nextWindow;
+
+					// doesn't propagate through onNext since window is unsent
+
+					previousState = commitWork(this, expectedState, true);
+					expectedState =
+							(((previousState &~ACTIVE_WINDOW_INDEX_MASK) &~ HAS_UNSENT_WINDOW) ^ (expectedState == previousState ? HAS_WORK_IN_PROGRESS : 0)) |
+									incrementActiveWindowIndex(previousState) |
+									HAS_UNSENT_WINDOW;
+					previousState |= HAS_UNSENT_WINDOW;
+
+					if (isCancelled(expectedState)) {
+						previousWindow.sendCancel();
+						nextWindow.sendCancel();
+						Operators.onDiscard(nextWindow, this.actual.currentContext());
```

Comment:
same as above, prefer triggering a discard of the queued values (which I think `sendCancel` should already cover)